### ISSUE TITLE
loop-vectorize introduces bad ptrtoint for non-integral address space

### ISSUE THREAD

**Issue by Keno** [llvm/llvm-project#54634]
Consider the following (reduced) .ll:
```
; ModuleID = 'reduced.ll'
source_filename = "reduced.ll"
target triple = "x86_64-unknown-linux-gnu"
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128-ni:10:11:12:13"

@SUM.CoreDOT.Tuple5334 = external dso_local global {}*
@jlplt_ijl_alloc_array_1d_10294_got = external dso_local global void ()*
@llvm.compiler.used = appending global [4 x i8*] [i8* bitcast (void ({} addrspace(10)*)* @ijl_gc_queue_root to i8*), i8* bitcast (void ({} addrspace(10)*)* @jl_gc_queue_binding to i8*), i8* bitcast ({} addrspace(10)* (i8*, i32, i32)* @ijl_gc_pool_alloc to i8*), i8* bitcast ({} addrspace(10)* (i8*, i64)* @ijl_gc_big_alloc to i8*)], section "llvm.metadata"

declare {}*** @julia.get_pgcstack()

define {} addrspace(10)* @japi1_vect_42283({} addrspace(10)** %0, i32 %1) #0 {
top:
  %2 = sext i32 %1 to i64
  %3 = load atomic {} addrspace(10)* ({} addrspace(10)*, i64)*, {} addrspace(10)* ({} addrspace(10)*, i64)** bitcast (void ()** @jlplt_ijl_alloc_array_1d_10294_got to {} addrspace(10)* ({} addrspace(10)*, i64)**) unordered, align 8
  %4 = call {} addrspace(10)* %3({} addrspace(10)* null, i64 0)
  br label %L26

L26:                                              ; preds = %L26, %top
  %value_phi5 = phi i64 [ 0, %top ], [ %value_phi8, %L26 ]
  %value_phi6 = phi i64 [ 0, %top ], [ %10, %L26 ]
  %5 = bitcast {} addrspace(10)** %0 to { {} addrspace(10)*, i64 } addrspace(10)**
  %6 = load { {} addrspace(10)*, i64 } addrspace(10)*, { {} addrspace(10)*, i64 } addrspace(10)** %5, align 8, !tbaa !0
  %7 = bitcast {} addrspace(10)* %4 to { {} addrspace(10)*, i64 } addrspace(13)* addrspace(10)*
  %8 = addrspacecast { {} addrspace(10)*, i64 } addrspace(13)* addrspace(10)* %7 to { {} addrspace(10)*, i64 } addrspace(13)* addrspace(11)*
  %9 = load { {} addrspace(10)*, i64 } addrspace(13)*, { {} addrspace(10)*, i64 } addrspace(13)* addrspace(11)* %8, align 8, !tbaa !5
  %.elt = getelementptr inbounds { {} addrspace(10)*, i64 }, { {} addrspace(10)*, i64 } addrspace(10)* %6, i64 0, i32 0
  %.unpack = load {} addrspace(10)*, {} addrspace(10)* addrspace(10)* %.elt, align 8, !tbaa !8
  %.elt1 = getelementptr inbounds { {} addrspace(10)*, i64 }, { {} addrspace(10)*, i64 } addrspace(10)* %6, i64 0, i32 1
  %.unpack2 = load i64, i64 addrspace(10)* %.elt1, align 8, !tbaa !8
  %.repack = getelementptr inbounds { {} addrspace(10)*, i64 }, { {} addrspace(10)*, i64 } addrspace(13)* %9, i64 %value_phi6, i32 0
  store {} addrspace(10)* %.unpack, {} addrspace(10)* addrspace(13)* %.repack, align 8, !tbaa !10
  %.repack4 = getelementptr inbounds { {} addrspace(10)*, i64 }, { {} addrspace(10)*, i64 } addrspace(13)* %9, i64 %value_phi6, i32 1
  store i64 %.unpack2, i64 addrspace(13)* %.repack4, align 8, !tbaa !10
  %10 = add i64 %value_phi6, 1
  %.not = icmp eq i64 %value_phi5, %2
  %11 = add i64 %value_phi5, 1
  %value_phi8 = select i1 %.not, i64 0, i64 %11
  br i1 %.not, label %L44, label %L26

L44:                                              ; preds = %L26
  ret {} addrspace(10)* null
}

; Function Attrs: inaccessiblemem_or_argmemonly
declare void @ijl_gc_queue_root({} addrspace(10)*) #1

; Function Attrs: inaccessiblemem_or_argmemonly
declare void @jl_gc_queue_binding({} addrspace(10)*) #1

; Function Attrs: allocsize(1)
declare noalias nonnull {} addrspace(10)* @ijl_gc_pool_alloc(i8*, i32, i32) #2

; Function Attrs: allocsize(1)
declare noalias nonnull {} addrspace(10)* @ijl_gc_big_alloc(i8*, i64) #2

attributes #0 = { "target-cpu"="skylake-avx512" "target-features"="+xsaves,+xsavec,+prfchw,+lzcnt,+sahf,+pku,+avx512vl,+avx512bw,+avx512cd,+clwb,+clflushopt,+adx,+avx512dq,+avx512f,+bmi2,+avx2,+bmi,+fsgsbase,+f16c,+avx,+xsave,+aes,+popcnt,+movbe,+sse4.2,+sse4.1,+cx16,+fma,+ssse3,+pclmul,+sse3,-rdrnd,-rtm,-rdseed,-avx512ifma,-avx512pf,-avx512er,-sha,-prefetchwt1,-avx512vbmi,-waitpkg,-avx512vbmi2,-shstk,-gfni,-vaes,-vpclmulqdq,-avx512vnni,-avx512bitalg,-avx512vpopcntdq,-rdpid,-cldemote,-movdiri,-movdir64b,-enqcmd,-avx512vp2intersect,-serialize,-tsxldtrk,-pconfig,-amx-bf16,-amx-tile,-amx-int8,-sse4a,-xop,-lwp,-fma4,-tbm,-mwaitx,-xsaveopt,-clzero,-wbnoinvd,-avx512bf16,-ptwrite,+sse2,+mmx,+fxsr,+64bit,+cx8" }
attributes #1 = { inaccessiblemem_or_argmemonly }
attributes #2 = { allocsize(1) }

!0 = !{!1, !1, i64 0}
!1 = !{!"jtbaa_value", !2, i64 0}
!2 = !{!"jtbaa_data", !3, i64 0}
!3 = !{!"jtbaa", !4, i64 0}
!4 = !{!"jtbaa"}
!5 = !{!6, !6, i64 0}
!6 = !{!"jtbaa_arrayptr", !7, i64 0}
!7 = !{!"jtbaa_array", !3, i64 0}
!8 = !{!9, !9, i64 0}
!9 = !{!"jtbaa_immut", !1, i64 0}
!10 = !{!11, !11, i64 0}
!11 = !{!"jtbaa_arraybuf", !2, i64 0}
```
On LLVM 14.0.0, running `opt -O3` introduces
```
  %77 = ptrtoint <4 x {} addrspace(10)*> %broadcast.splat9 to <4 x i64>
  %interleaved.vec12.epil = shufflevector <4 x i64> %77, <4 x i64> %broadcast.splat11, <8 x i32> <i32 0, i32 4, i32 1, i32 5, i32 2, i32 6, i32 3, i32 7>
  store <8 x i64> %interleaved.vec12.epil, <8 x i64> addrspace(13)* %71, align 8, !tbaa !10
```
The `ptrtoint` here is not legal, because addrspace 10 is non-integral.

**Comment by preames** [comment]
@keno, I assume you're reproducing the above with a build with assertions disabled right?

If I run the above example through either ToT opt or LLVM-14 opt, I get the following assertion failure:
$ ./opt -loop-vectorize pr54634-lv.ll -S -enable-interleaved-mem-accesses=1
opt: /home/ubuntu/llvm-repo/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:2971: llvm::Value* llvm::InnerLoopVectorizer::createBitOrPointerCast(llvm::Value*, llvm::VectorType*, const llvm::DataLayout&): Assertion `(DstElemTy->isFloatingPointTy() != SrcElemTy->isFloatingPointTy()) && "Only one type should be a floating point type"' failed.

**Comment by Keno** [comment]
Ah, yes, sorry the reproducer was run without LLVM assertions enabled. If I enable assertions, I see the same assertion you report.

**Comment by preames** [comment]
Should be fixed by commit 88de27e3

**Comment by Keno** [comment]
Confirmed to fix the original case this was reduced from. Thanks!
